### PR TITLE
Create Quirk for TS011F

### PR DESCRIPTION
**Pull request by miit86**
Quirk for the TS011F 
See Issue https://github.com/zigpy/zha-device-handlers/issues/1045

### CODE REVIEW

**Comment by miit86** [comment]
My bad,

I saw the mistake. and corrected it right away. 
Still having the unknown cluster warning in logs. can we do something about that?

**Comment by MattWestb** [comment]
Try deleting the:
```
                    0xE000,  # Unknown
                    0xE001,  # Unknown
```
in the replace part, i think then ZHA is not getting the clusters and they is going to "dev nul".

**Comment by miit86** [comment]
I removed
```
                    0xE000,  # Unknown
                    0xE001,  # Unknown
```
But still same entry in logs. But Im not sure if there related.
have more tuya stuff.

**Comment by MattWestb** [comment]
You can trying defining it like this (from IKEA Styrbar):
https://github.com/zigpy/zha-device-handlers/blob/ade14754c19a4524b22583ce37fd61bcbc863ad6/zhaquirks/ikea/fourbtnremote.py#L46
but 2 different name for the clusters.

Perhaps David is knowing how to do it OK ?

File: zhaquirks/tuya/TS011F_plug.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    INPUT_CLUSTERS,
+    MODEL,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+)
+
+class TuyaZBMetering(CustomCluster, Metering):
+    """TuyaZBMetering cluster implementation. This implementation divides the KwH for tuya devices"""
+
+    cluster_id = Metering.cluster_id
```
Comment (Adminiuga) [inline zhaquirks/tuya/TS011F_plug.py:21]:
```suggestion
```

No need to redefine this

File: zhaquirks/tuya/TS011F_plug.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    current_summ_delivered = 0x0000
+
+    def _update_attribute(self, attrid, value):
+        if attrid == self.current_summ_delivered:
+            value = value / 100
+        super()._update_attribute(attrid, value)
```
Comment (Adminiuga) [inline zhaquirks/tuya/TS011F_plug.py:27]: Check https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/edpwithus/__init__.py Can you test with 
```suggestion
    MULTIPLIER = 0x0301
    DIVISOR = 0x0302
    _CONSTANT_ATTRIBUTES = {MULTIPLIER: 1, DIVISOR: 100}
```

File: zhaquirks/tuya/TS011F_plug.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class TuyaZBElectricalMeasurement(CustomCluster, ElectricalMeasurement):
+    """TuyaZBElectricalMeasurement cluster implementation. This implementation divides the Current for tuya devices"""
+
+    cluster_id = ElectricalMeasurement.cluster_id
+    rms_current = 0x0508
+
+    def _update_attribute(self, attrid, value):
+        if attrid == self.rms_current:
+            value = value / 1000
+        super()._update_attribute(attrid, value)
```
Comment (Adminiuga) [inline zhaquirks/tuya/TS011F_plug.py:38]:
```suggestion
    AC_VOLTAGE_MULTIPLIER = 0x0600
    AC_VOLTAGE_DIVISOR = 0x0601
    
    _CONSTANT_ATTRIBUTES = {AC_VOLTAGE_MULTIPLIER: 1, AC_VOLTAGE_DIVISOR: 100}
```

Try just defining the divisor and multiplier

File: zhaquirks/tuya/TS011F_plug.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Off = 0x00
+    On = 0x01
+    LastState = 0x02
+
+
+class OnOffRestorePowerCluster(CustomCluster, OnOff):
+    """Tuya on off cluster with restore state."""
+
+    attributes = OnOff.attributes.copy()
+    attributes.update({0x8002: ("power_on_state", PowerOnState)})
```
Comment (Adminiuga) [inline zhaquirks/tuya/TS011F_plug.py:52]:
```suggestion
    manufacturer_attributes = {0x8002: ("power_on_state", PowerOnState)}
```

**Review by Adminiuga — CHANGES_REQUESTED** [review]
Run through black and isort to fix formatitng.
Test and consider the proposed changes
Otherwise looks good

File: zhaquirks/tuya/TS011F_plug.py (hunk 1 above)
Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py:21]: This is OK

also removed line 32: cluster_id = ElectricalMeasurement.cluster_id

File: zhaquirks/tuya/TS011F_plug.py (hunk 2 above)
Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py:27]: Hi, not working anymore. Value isn't divided by 100 anymore

File: zhaquirks/tuya/TS011F_plug.py (hunk 3 above)
Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py:38]: Hi, not working anymore. Value isn't divided by 1000 anymore
AC_Voltage_Divisor should be 1000

File: zhaquirks/tuya/TS011F_plug.py (hunk 4 above)
Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py:52]: Not working:
PowerOnState Value is always None
But I know matt has a PR for implementing this in INIT

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py:52]: Then its very likely not working for the moment then i have doing the same change in the tuya INIT.

I can rolling it back for getting it working if its OK for Adminiuga in the PR for it.

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py:52]: PS the devs is in time zone US so they is having there night sleep for the moment if not doing dream coding ;-))

**Comment by MattWestb** [comment]
If you is adding this at line 25 with one empty line before and 2 after:
```
TUYA_CLUSTER_ID0 = 0xE000 # 57344
TUYA_CLUSTER_ID1 = 0xE001 # 57345
```
and changing the line 79 and 80 to this (Signature cluster 0xE000 and 0xE001) :
```
                    TUYA_CLUSTER_ID0
                    TUYA_CLUSTER_ID1
```
and deleting the line 109 and 110 (in the replacement):
```
                    0xE000,  # Unknown
                    0xE001,  # Unknown
```
Do you still getting warning in HA long from the device ?

File: zhaquirks/tuya/TS011F_plug.py (hunk 4 above)
Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py]: Using TuyaZBOnOffRestorePowerCluster from INIT: PR1056

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: Only need getting it merged so you is not getting errors in the tests (i have it in the old plug.py then deleting the original code).

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: You need adding some `.` in the end of your new comments and running black for fixing some formatting that is not OK (then looking on the last test done).

Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py]: black for formatting, I thought I did :)
where do I need to add ```.``` ?

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: Line 29 and 37.
Its some missing blank lines and extra spaces but that is black fixing that also some to long strings that is being broken in up OK by black (or its being shorted then deleting the power restore).

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: the last test run: https://github.com/zigpy/zha-device-handlers/pull/1057/checks?check_run_id=3785501424 look in `check flake8`

Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py]: I don't use this part in the code anymore:
```
attributes = OnOff.attributes.copy()
attributes.update({0x8002: ("power_on_state", PowerOnState)})
```

and a lot has changed: 
Need a new test.

Comment (miit86) [reply zhaquirks/tuya/TS011F_plug.py]: Hi Matt, Did I forgot something? or how to go on from here?

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: Still line 29 and 37 need being ended with one  `.`  (like line 1 and 45 is done) that black is not fixing the rest is black fixing then you running it.
You is also getting one (or more) error for line 25 then the power on state is not committed but its OK then the PR is merged.

Comment (MattWestb) [reply zhaquirks/tuya/TS011F_plug.py]: Looks like all lines is OK :-)))

Hope the other test is not finding any bad things and that the devs can rerunning the test and merger it if its OK :-)))

**Comment by miit86** [comment]
Whats Wrong With ISORT ?

**Comment by MattWestb** [comment]
ISORT like having the block ` from zhaquirks.const import (` after line 25 (the last import then its the last in the sorting) and before ` class TuyaZBMetering(CustomCluster, Metering):`  ;-(((

**Comment by MattWestb** [comment]
The tests is one large PITA but needed.

I normally looking on other quirks and then always using black playground (it fixing most bad things but not all sorting) but in the end its the gut tests that must working.

Hope Denis can redoing the test then you have making the change and getting it merged !!

**Comment by MattWestb** [comment]
You need moving the 
```
from zhaquirks.const import (
    DEVICE_TYPE,
    ENDPOINTS,
    INPUT_CLUSTERS,
    MODEL,
    OUTPUT_CLUSTERS,
    PROFILE_ID,
)
```
to the same possition.
From the test:
```
Deleting:

-from zhaquirks.const import (
-    DEVICE_TYPE,
-    ENDPOINTS,
-    INPUT_CLUSTERS,
-    MODEL,
-    OUTPUT_CLUSTERS,
-    PROFILE_ID,
-)
-from zhaquirks.tuya import TuyaZBOnOffRestorePowerCluster

Adding between this lines:

 from zigpy.zcl.clusters.homeautomation import ElectricalMeasurement
 from zigpy.zcl.clusters.smartenergy import Metering
 
+from zhaquirks.const import (
+    DEVICE_TYPE,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    MODEL,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+)
+from zhaquirks.tuya import TuyaZBOnOffRestorePowerCluster
+
 
 class TuyaZBMetering(CustomCluster, Metering):
     """Divides the KwH for tuya."""
````

**Comment by miit86** [comment]
no it's from https://pycqa.github.io/isort/docs/quick_start/0.-try.html 

and the H is before the I

**Comment by miit86** [comment]
@MattWestb 
Fingers Crossed..!

**Comment by MattWestb** [comment]
THAT looks being very OK !!

I do all my fingers and also toes :-)))

**Comment by MattWestb** [comment]
Make it like in https://github.com/zigpy/zha-device-handlers/pull/1057#issuecomment-933593639 but updating the lines then they can being changed.
And 2 ways doing the replacement. 1 using the name defined, 2 deleting them so the system is not using them.

**Comment by miit86** [comment]
@MattWestb 
AAARGGH, same error. because it can't find the clusters.

maybe I should define it somewhere in INIT maybe.

```
class TUYA_CLUSTER_ID0(CustomCluster):
    """Tuya manufacturer specific cluster 57344."""

    name = "Tuya Manufacturer Specific"
    cluster_id = TUYA_CLUSTER_ID0


class TUYA_CLUSTER_ID1(CustomCluster):
    """Tuya manufacturer specific cluster 57345."""

    name = "Tuya Manufacturer Specific"
    cluster_id = TUYA_CLUSTER_ID1
```

also in "ts0201_zemismart.py" they use a: ```0xEEFF,  # Unknown```
```
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.TEMPERATURE_SENSOR,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    PowerConfiguration.cluster_id,
                    Identify.cluster_id,
                    RelativeHumidity.cluster_id,
                    TemperatureMeasurement.cluster_id,
                    0xEEFF,  # Unknown
                ],
                OUTPUT_CLUSTERS: [Identify.cluster_id, Ota.cluster_id],
            }
        },
```


@Adminiuga 
can you Help? My knowledge of coding isn't that good.

**Comment by MattWestb** [comment]
Deleting line 27,28 and 29 (they is now not needed).
Make line 85 and 86 like line 115 and 166 then its looks nicer.

The old PITA was Xiaomi and the new is tuya !!!

I think its looks very good (without any code knowledge at all  !!).

**Comment by MattWestb** [comment]
Ops Black like have little more spaces in 2 lines ;-((
```
                     OnOff.cluster_id,
                     Metering.cluster_id,
                     ElectricalMeasurement.cluster_id,
-                    0xE000, # TuyaClusterE000
-                    0xE001, # TuyaClusterE001
+                    0xE000,  # TuyaClusterE000
+                    0xE001,  # TuyaClusterE001
                 ],
                 OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
             },
```

**Comment by MattWestb** [comment]
Then you is getting it loading do you getting the warning you was getting before ?
I think they shall being fixed with the right config of the cluster in the replacement.

**Comment by MattWestb** [comment]
Than all is working OK is the devs merging it to dev branch and you can using it if you like and then its being released in the next HA release if all is going well.

For using it before its in HA you can installing by configuring local quirks in HA and adding the new quirk in it https://github.com/zigpy/zha-device-handlers/discussions/693#discussioncomment-457020. 
The problem is that this PR is needing https://github.com/zigpy/zha-device-handlers/pull/1056 changed in the tuya INI so it not working and you need patching the ZHA-quirks in your HA docker container as in this guide:https://github.com/zigpy/zha-device-handlers/discussions/693#discussion-1237451 both the tuya INI and the new ts011f.py need being updated for getting it working.

If  you is god in coding you can doing one version with the power state and dont need the tuya INI but its little more work.

**Comment by miit86** [comment]
@lemppari 

1. enable custom quirks in HA
2. copy code below to new file TS011F.plug 
3. remove device from ZHA and Add again.
4. Reset the Device if you want to start with fresh.

```
"""TS011F plug."""
import zigpy.types as t
from zhaquirks.const import (
    DEVICE_TYPE,
    ENDPOINTS,
    INPUT_CLUSTERS,
    MODEL,
    OUTPUT_CLUSTERS,
    PROFILE_ID,
)
from zigpy.profiles import zha
from zigpy.quirks import CustomCluster, CustomDevice
from zigpy.zcl.clusters.general import (
    Basic,
    GreenPowerProxy,
    Groups,
    Identify,
    OnOff,
    Ota,
    Scenes,
    Time,
)
from zigpy.zcl.clusters.homeautomation import ElectricalMeasurement
from zigpy.zcl.clusters.smartenergy import Metering


class TuyaClusterE000(CustomCluster):
    """Tuya manufacturer specific cluster 57344."""

    name = "Tuya Manufacturer Specific"
    cluster_id = 0xE000
    ep_attribute = "tuya_is_pita_0"


class TuyaClusterE001(CustomCluster):
    """Tuya manufacturer specific cluster 57345."""

    name = "Tuya Manufacturer Specific"
    cluster_id = 0xE001
    ep_attribute = "tuya_is_pita_1"


class TuyaZBMetering(CustomCluster, Metering):
    """TuyaZBMetering cluster divides the KwH for tuya"""

    MULTIPLIER = 0x0301
    DIVISOR = 0x0302
    _CONSTANT_ATTRIBUTES = {MULTIPLIER: 1, DIVISOR: 100}


class TuyaZBElectricalMeasurement(CustomCluster, ElectricalMeasurement):
    """TuyaZBElectricalMeasurement cluster divides the Current for tuya"""

    AC_VOLTAGE_MULTIPLIER = 0x0600
    AC_VOLTAGE_DIVISOR = 0x0601
    _CONSTANT_ATTRIBUTES = {AC_VOLTAGE_MULTIPLIER: 1, AC_VOLTAGE_DIVISOR: 1000}


class PowerOnState(t.enum8):
    """Tuya power on state enum."""

    Off = 0x00
    On = 0x01
    LastState = 0x02


class OnOffRestorePowerCluster(CustomCluster, OnOff):
    """Tuya on off cluster with restore state."""

    attributes = OnOff.attributes.copy()
    attributes.update({0x8002: ("power_on_state", PowerOnState)})


class Plug(CustomDevice):
    """Tuya plug with restore power state support."""

    signature = {
        MODEL: "TS011F",
        ENDPOINTS: {
            # <SimpleDescriptor endpoint=1 profile=266 device_type=81
            # device_version=1
            # input_clusters=[0, 3, 4, 5, 6, 9, 1794, 2820, 57344, 57345]
            # output_clusters=[10, 25]>
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.ON_OFF_PLUG_IN_UNIT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Identify.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOff.cluster_id,
                    Metering.cluster_id,
                    ElectricalMeasurement.cluster_id,
                    TuyaClusterE000.cluster_id,
                    TuyaClusterE001.cluster_id,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
            # <SimpleDescriptor endpoint=242 profile=41440 device_type=97
            # device_version=1
            # input_clusters=[]
            # output_clusters=[33]>
            242: {
                PROFILE_ID: 41440,
                DEVICE_TYPE: 97,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }
    replacement = {
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.ON_OFF_PLUG_IN_UNIT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Identify.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOffRestorePowerCluster,
                    TuyaZBMetering,
                    TuyaZBElectricalMeasurement,
                    TuyaClusterE000,
                    TuyaClusterE001,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
        },
    }

```

**Comment by miit86** [comment]
YEEEH..! Thanks Guys

**Pull request merged**